File: code/algorithms/strings/LongestCommonPrefixArray.py


class SuffixArray:
    def __init__(self, text):
        self.ALPHABET_SZ = 256
        self.T = text
        self.N = len(text)
        self.sa = [0] * self.N
        self.sa2 = [0] * self.N
        self.rank = [0] * self.N
        self.c = [0] * max(self.ALPHABET_SZ, self.N)
        self.construct()
        self.kasai()

    def construct(self):
        c = [0] * self.ALPHABET_SZ
        for i in range(self.N):
            self.rank[i] = self.T[i]
            c[self.rank[i]] += 1
        for i in range(1, self.ALPHABET_SZ):
            c[i] += c[i - 1]
        for i in range(self.N - 1, -1, -1):
            c[self.T[i]] -= 1
            self.sa[c[self.T[i]]] = i
        p = r = 0
        while p < self.N:
            r = 0
            for i in range(self.N - p, self.N):
                self.sa2[r] = i
                r += 1
            for i in range(self.N):
                if self.sa[i] >= p:
                    self.sa2[r] = self.sa[i] - p
                    r += 1
            c = [0] * self.ALPHABET_SZ
            for i in range(self.N):
                c[self.rank[i]] += 1
            for i in range(1, self.ALPHABET_SZ):
                c[i] += c[i - 1]
            for i in range(self.N - 1, -1, -1):
                c[self.rank[self.sa2[i]]] -= 1
                self.sa[c[self.rank[self.sa2[i]]]] = self.sa2[i]
            self.sa2[self.sa[0]] = r = 0
            for i in range(1, self.N):
                if not (self.rank[self.sa[i - 1]] == self.rank[self.sa[i]] and
                        self.sa[i - 1] + p < self.N and self.sa[i] + p < self.N and
                        self.rank[self.sa[i - 1] + p] == self.rank[self.sa[i] + p]):
                    r += 1
                self.sa2[self.sa[i]] = r
            self.rank, self.sa2 = self.sa2, self.rank
            if r == self.N - 1:
                break
            self.ALPHABET_SZ = r + 1
            p <<= 1

    def kasai(self):
        self.lcp = [0] * self.N
        inv = [0] * self.N
        for i in range(self.N):
            inv[self.sa[i]] = i
        i = 0
        while i < self.N:
            if inv[i] > 0:
                k = self.sa[inv[i] - 1]
                length = 0
                while i + length < self.N and k + length < self.N and self.T[i + length] == self.T[k + length]:
                    length += 1
                self.lcp[inv[i] - 1] = length
                if length > 0:
                    length -= 1
            i += 1

    def display(self):
        print("-----i-----SA-----LCP---Suffix")
        for i in range(self.N):
            suffix_len = self.N - self.sa[i]
            suffix = ''.join(chr(c) for c in self.T[self.sa[i]:])
            print(f"{i:7} {self.sa[i]:7} {self.lcp[i]:7} {suffix}")


if __name__ == "__main__":
    str_ = "abcbbccaabcaabc"
    sa = SuffixArray([ord(c) for c in str_])
    sa.display()
    print("The LCP array is:", sa.lcp)

File: code/algorithms/strings/LongestCommonPrefixArray.cpp


#include <iostream>
#include <vector>
#include <algorithm>

using namespace std;

class SuffixArray {
public:
    int ALPHABET_SZ = 256;
    int N;
    vector<int> T, lcp, sa, sa2, rank, tmp, c;

    SuffixArray(string str) {
        T = toIntArray(str);
        N = str.length();
        sa.resize(N);
        sa2.resize(N);
        rank.resize(N);
        c.resize(max(ALPHABET_SZ, N));
        construct();
        kasai();
    }

    vector<int> toIntArray(string s) {
        vector<int> text(s.length());
        for (int i = 0; i < s.length(); i++) text[i] = s[i];
        return text;
    }

    void construct() {
        vector<int> c(ALPHABET_SZ);
        for (int i = 0; i < N; ++i) c[rank[i] = T[i]]++;
        for (int i = 1; i < ALPHABET_SZ; ++i) c[i] += c[i - 1];
        for (int i = N - 1; i >= 0; --i) sa[--c[T[i]]] = i;
        for (int p = 1, r = 0; p < N; p <<= 1) {
            for (int i = N - p, r = 0; i < N; ++i) sa2[r++] = i;
            for (int i = 0; i < N; ++i) if (sa[i] >= p) sa2[r++] = sa[i] - p;
            fill(c.begin(), c.begin() + ALPHABET_SZ, 0);
            for (int i = 0; i < N; ++i) c[rank[i]]++;
            for (int i = 1; i < ALPHABET_SZ; ++i) c[i] += c[i - 1];
            for (int i = N - 1; i >= 0; --i) sa[--c[rank[sa2[i]]]] = sa2[i];
            int i;
            for (sa2[sa[0]] = r = i = 0; i < N; ++i) {
                if (!(rank[sa[i - 1]] == rank[sa[i]]
                        && sa[i - 1] + p < N && sa[i] + p < N
                        && rank[sa[i - 1] + p] == rank[sa[i] + p])) r++;
                sa2[sa[i]] = r;
            }
            tmp = rank;
            rank = sa2;
            sa2 = tmp;
            if (r == N - 1) break;
            ALPHABET_SZ = r + 1;
        }
    }

    void kasai() {
        lcp.resize(N);
        vector<int> inv(N);
        for (int i = 0; i < N; i++) inv[sa[i]] = i;
        for (int i = 0, len = 0; i < N; i++) {
            if (inv[i] > 0) {
                int k = sa[inv[i] - 1];
                while (i + len < N && k + len < N && T[i + len] == T[k + len]) len++;
                lcp[inv[i] - 1] = len;
                if (len > 0) len--;
            }
        }
    }

    void display() {
        cout << "-----i-----SA-----LCP---Suffix" << endl;
        for (int i = 0; i < N; i++) {
            int suffixLen = N - sa[i];
            string suffix(T.begin() + sa[i], T.end());
            cout << i << "       " << sa[i] << "      " << lcp[i] << "    " << suffix << endl;
        }
    }
};

int main() {
    string str = "abcbbccaabcaabc";
    SuffixArray sa(str);
    sa.display();
    cout << "The LCP array is: ";
    for (int i = 0; i < sa.lcp.size(); ++i) {
        cout << sa.lcp[i] << " ";
    }
    cout << endl;
    return 0;
}
